ksql-engine/src/test/java/io/confluent/ksql/integration/EndToEndIntegrationTest.java

```diff
@@ -0,0 +1,296 @@
+
+    Set<String> actualUsers = new HashSet<>();
+    Set<String> expectedUsers = Utils.mkSet("USER_0", "USER_1", "USER_2", "USER_3", "USER_4");
+    while (actualUsers.size() < 5) {
+      KeyValue<String, GenericRow> nextRow = rowQueue.poll();
+      if (nextRow != null) {
+        List<Object> columns = nextRow.value.getColumns();
+        assertEquals(6, columns.size());
+        actualUsers.add((String) columns.get(1));
+      }
+    }
+    assertEquals(expectedUsers, actualUsers);
+    queryMetadata.getKafkaStreams().close();
+  }
+
+  private void validateSelectFromPageViewsWithSpecificColumn() throws Exception {
+    String query = String.format("SELECT pageid from %s;", pageViewStream);
+    log.debug("Sending query: {}", query);
+
+    List<QueryMetadata> queries = ksqlEngine.buildMultipleQueries(false, query, Collections.emptyMap());
+
+    assertEquals(1, queries.size());
+    assertTrue(queries.get(0) instanceof QueuedQueryMetadata);
+
+    QueuedQueryMetadata queryMetadata = (QueuedQueryMetadata) queries.get(0);
+    queryMetadata.getKafkaStreams().start();
+
+    BlockingQueue<KeyValue<String, GenericRow>> rowQueue = queryMetadata.getRowQueue();
+
+    List<String> actualPages = new ArrayList<>();
+    List<String> expectedPages = Arrays.asList("PAGE_1", "PAGE_2", "PAGE_3", "PAGE_4", "PAGE_5", "PAGE_5", "PAGE_5");
+    while (actualPages.size() < expectedPages.size()) {
+      KeyValue<String, GenericRow> nextRow = rowQueue.poll();
+      if (nextRow != null) {
+        List<Object> columns = nextRow.value.getColumns();
+        assertEquals(1, columns.size());
+        String page = (String) columns.get(0);
+        actualPages.add(page);
+      }
+    }
+
+    assertEquals(expectedPages, actualPages);
+    queryMetadata.getKafkaStreams().close();
+  }
+
+  private String validateSelectAllFromDerivedStream() throws Exception {
+
+    String derivedStream = "pageviews_female";
+    createPageViewsFemaleStream(derivedStream);
+
+    String selectQuery = String.format("SELECT * from %s;", derivedStream);
+
+    List<QueryMetadata> queries = ksqlEngine.buildMultipleQueries(false, selectQuery, Collections.emptyMap());
+
+    assertEquals(1, queries.size());
+    assertTrue(queries.get(0) instanceof QueuedQueryMetadata);
+
+    QueuedQueryMetadata queryMetadata = (QueuedQueryMetadata) queries.get(0);
+    queryMetadata.getKafkaStreams().start();
+
+    List<KeyValue<String, GenericRow>> results = new ArrayList<>();
+    BlockingQueue<KeyValue<String, GenericRow>> rowQueue = queryMetadata.getRowQueue();
+
+    // From the mock data, we expect exactly 3 page views from female users.
+
+    List<String> expectedPages = Arrays.asList("PAGE_2", "PAGE_5", "PAGE_5");
+    List<String> expectedUsers = Arrays.asList("USER_2", "USER_0", "USER_2");
+    List<String> actualPages = new ArrayList<>();
+    List<String> actualUsers = new ArrayList<>();
+
+    while (results.size() < 3) {
```

Comment:
Do we need a timeout on this loop? Or maybe use `TestUtils#waitForCondition`? i.e., is there a chance this will loop forever?

Reply:
Yea. I was thinking about that. It's better to just fail out if we don't get anything after a minute or something, rather than having the test hang.